docs/overwatch_docs/content/UserGuide/Configuration.md
---
title: "Configuration"
date: 2020-10-28T11:55:12-04:00
draft: true
weight: 2
---

The configuration for Overwatch is fairly basic for the initial release but the plan is to open a lot of options
and make the config much more robust to allow for much more granular control over operations. The ongoing changes
will be documented here.

## Configuration Basics
The Overwatch configuration can be created as a case class of `OverwatchParams` or as a json string passed into
the main class `com.databricks.labs.overwatch.BatchRunner`. When passed in as a json string, it is 
deserialized into an instance of `OverwatchParams`. This provides strong validation on the input parameters
and strong typing for additional validation options. All configs attempt to be defaulted to a reasonable default
where possible. If a default is set, passing in a value will simply overwrite the default.

## Azure Example Config

## AWS Example Config

### Simplified Example
```scala
private val db = "Overwatch_Test"
private val dataTarget = DataTarget(Some(db), Some(s"dbfs:/user/hive/warehouse/${db}.db"))
private val tokenSecret = TokenSecret("db_services", "databricks_overwatch")
private val overwatchModules = "audit,accounts,sparkEvents,jobs,clusters,clusterEvents,notebooks".split(",")

val params = OverwatchParams()
```

### Real-world Example
```scala
private val db = "Overwatch_Test"
private val dataTarget = DataTarget(Some(db), Some(s"dbfs:/user/hive/warehouse/${db}.db"))
private val tokenSecret = TokenSecret("db_services", "databricks_overwatch")
private val overwatchModules = "audit,accounts,sparkEvents,jobs,clusters,clusterEvents,notebooks".split(",")

val params = OverwatchParams(
  auditLogConfig = AuditLogConfig(Some("s3a://CORP_audit_logBucket/databricks/audit-logs")),
  tokenSecret = Some(tokenSecret),
  dataTarget = Some(dataTarget), 
  badRecordsPath = Some(s"/tmp/overwatch/sparkEventsBadrecords"),
  overwatchScope = Some(overwatchModules),
  maxDaysToLoad = 30
)
```

### OverwatchParams
The configuration structure required for Overwatch

Config | Required Override | Default Value | Type | Description
:--------------------------|:---|:----------|:----------|:--------------------------------------------------
**auditLogConfig**|Y|NULL|[AuditLogConfig](#auditlogconfig)|Databricks Audit Log delivery information.
**tokenSecret**|N|Execution Owner's Token|Option[TokenSecret](#tokensecret)|Secret retrieval information
**dataTarget**|N|[DataTarget](#datatarget)|Option[DataTarget](#datatarget)|What to call the database and where to store it
**badRecordsPath**|N|/tmp/overwatch/badRecordsPath|Option[String]|When reading the log files, where should Overwatch store the records / files that cannot be parsed. Overwatch must have write permissions to this path 
**overwatchScope**|N|all|Option[Seq[String]]|List of modules in scope for the run. It's important to note that there are many co-dependencies. When choosing a module, be sure to also enable it's requisites. If not value provided, all modules will execute.
**maxDaysToLoad**|N|60|Int|On large, busy workspaces 60 days of data may amount in 10s of TB of raw data. This parameter allows the job to be broken out into several smaller runs.

### AuditLogConfig
Config to point Overwatch to the location of the delivered audit logs

Config | Required Override | Default Value | Type | Description
:--------------------------|:---|:----------|:----------|:--------------------------------------------------
**rawAuditPath**|Y|None|Option[String]|Top-level path to directory containing workspace audit logs delivered by Databricks. The Overwatch user must have read access to this path
**azureAuditLogEventhubConfig**|Y (only on Azure)|None|Option[AzureAuditLogEventhubConfig](#azureauditlogeventhubconfig)|Required configuration when using Azure as Azure must deliver audit logs via LogAnalytics via Eventhub

### TokenSecret
Overwatch must have permission to perform its functions; these are further discussed in [AdvancedTopics](advancedtopics.md).
The token secret stores the Databricks Secret scope / key for Overwatch to retrieve. The key should store the 
token secret to be used which usually starts with "dapi..." 

If no TokenSecret is passed into the config, the operation owner's token will be used. If Overwatch is being 
executed in a notebook the notebook user's token will be used. If Overwatch is being executed through a job the 
token of the job owner will be used. Whatever token is used, it must have the appropriate access or it will result 
in missing data, or an Overwatch run failure.

Config | Required Override | Default Value | Type | Description
:--------------------------|:---|:----------|:----------|:--------------------------------------------------
**scope**|N|NA|String|Databricks secret scope 
**key**|N|NA|String|Databricks secret key within the scope defined in the Scope parameter

### DataTarget
Where to create the database and what to call it. This must be defined on first run or Overwatch will create a
database named "Overwatch" at the default location which is "/user/hive/warehouse/overwatch.db". This is challenging
to change later, so be sure you choose a good starting point. After the initial run, this must not change without 
and entire [database migration](AdvancedTopics.md). Overwatch performs destructive tasks within its own database and
this is how it protects itself against harming existing data. Overwatch creates specific metadata inside the database
at creation time to ensure the database is created and owned by the Overwatch process. Furthermore, metadata is 
managed to track schema versions and other states.

Config | Required Override | Default Value | Type | Description
:--------------------------|:---|:----------|:----------|:--------------------------------------------------
**databaseName**|N|overwatch|Option[String]|Name of database to be created on first run or to be appended on subsequent runs
**databaseLocation**|N|/user/hive/warehouse/{databaseName}.db|Option[String]|Location of the Overwatch database. Any compatible fully-qualified URI can be used here as long as Overwatch has access to write the target. Most customers, however, mount the qualified path and reference the mount point for simplicity but this is not required and may not be possible depending on security requirements and environment. 

### AzureAuditLogEventhubConfig
Not Required when using AWS <br>
Eventhub streaming environment configurations 

Config | Required Override | Default Value | Type | Description
:--------------------------|:---|:----------|:----------|:--------------------------------------------------
**connectionString**|Y|NA|String|Retrieve from Azure Portal Event Hub
**eventHubName**|Y|NA|String|Retrieve from Azure Portal Event Hub
**auditRawEventsPrefix**|Y|NA|String|Path prefix for checkpoint directories
**maxEventsPerTrigger**|N|10000|Int|Events to pull for each trigger, this should be increased during initial cold runs or runs that have very large numbers of audit log events.
**auditRawEventsChk**|N|{auditRawEventsPrefix}/rawEventsCheckpoint|Option[String]|Checkpoint Directory name for the raw dump of events from Eventhub. This directory gets overwritten upon successful pull into Overwatch.
**auditLogChk**|N|{auditRawEventsPrefix}/auditLogBronzeCheckpoint|Option[String]|**DEPRECATED** Checkpoint Directory name for the audit log stream target. This target will continuously grow as more audit logs are created and delivered


src/main/scala/com/databricks/labs/overwatch/utils/Config.scala
package com.databricks.labs.overwatch.utils

import java.time.temporal.ChronoUnit
import java.time.{Instant, LocalDateTime, ZoneId, ZoneOffset}
import java.util.{Calendar, Date, TimeZone, UUID}

import com.databricks.dbutils_v1.DBUtilsHolder.dbutils
import org.apache.log4j.{Level, Logger}
import org.apache.spark.sql.functions.{from_unixtime, lit}

class Config() {

  private final val _overwatchSchemaVersion = "0.1"
  private final val _runID = UUID.randomUUID().toString.replace("-","")
  private final val cipherKey = UUID.randomUUID().toString
  private val _isLocalTesting: Boolean = System.getenv("OVERWATCH") == "LOCAL"
  private val _isDBConnect: Boolean = System.getenv("DBCONNECT") == "TRUE"
  private var _isFirstRun: Boolean = false
  private var _debugFlag: Boolean = false
  private var _lastRunDetail: Array[ModuleStatusReport] = Array[ModuleStatusReport]()
  private var _pipelineSnapTime: Long = _
  private var _databaseName: String = _
  private var _databaseLocation: String = _
  private var _workspaceUrl: String = _
  private var _cloudProvider: String = _
  private var _token: Array[Byte] = _
  private var _tokenType: String = _
  private var _apiEnv: ApiEnv = _
  private var _auditLogConfig: AuditLogConfig = _
  private var _badRecordsPath: String = _
  private var _maxDays: Int = 60
  private var _passthroughLogPath: Option[String] = None
  private var _inputConfig: OverwatchParams = _
  private var _parsedConfig: ParsedConfig = _
  private var _overwatchScope: Seq[OverwatchScope.Value] = OverwatchScope.values.toSeq
  private var _initialSparkConf: Map[String, String] = Map()
  private var _intialShuffleParts: Int = 200

  final private val cipher = new Cipher(cipherKey)
  private val logger: Logger = Logger.getLogger(this.getClass)

  /**
   * Absolute oldest date for which to pull data. This is to help limit the stress on a cold start / gap start.
   * If trying to pull more than 60 days of data before https://databricks.atlassian.net/browse/SC-38627 is complete
   * The primary concern is that the historical data from the cluster events API generally expires on/before 60 days
   * and the event logs are not stored in an optimal way at all. SC-38627 should help with this but for now, max
   * age == 60 days.
   * @return
   */
  private def primordealEpoch: Long = {
    LocalDateTime.now(Config.utcZone).minusDays(60).toInstant(ZoneOffset.UTC)
      .truncatedTo(ChronoUnit.DAYS)
      .toEpochMilli
  }

  /**
   * This is also used for simulation of start/end times during testing. This also should not be a public function
   * when completed. Note that this controlled by module. Not every module is executed on every run or a module could
   * fail and this allows for missed data since the last successful run to be acquired without having to pull all the
   * data for all modules each time.
   * @param moduleID moduleID for modules in scope for the run
   * @throws java.util.NoSuchElementException
   * @return
   */
  @throws(classOf[NoSuchElementException])
  def fromTime(moduleID: Int): TimeTypes = {
    val lastRunStatus = if (!isFirstRun)
      lastRunDetail.filter(_.moduleID == moduleID)
    else
      lastRunDetail
    require(lastRunStatus.length <= 1, "More than one start time identified from pipeline_report.")
    val fromTime = if (lastRunStatus.length != 1)
      primordealEpoch
    else
      lastRunStatus.head.untilTS
    Config.createTimeDetail(fromTime)
  }

  private[overwatch] def daysToProcess(moduleId: Int): Long = {
    ChronoUnit.DAYS.between(fromTime(moduleId).asLocalDateTime, untilTime(moduleId).asLocalDateTime)
  }

  /**
   * BEGIN GETTERS
   * The next section is getters that provide access to local configuration variables. Only adding details where
   * the getter may be obscure or more complicated.
   */

  /**
   * Getter for Pipeline Snap Time
   * NOTE: PipelineSnapTime is EXCLUSIVE meaning < ONLY NOT <=
   * @return
   */
  def pipelineSnapTime: TimeTypes = {
    Config.createTimeDetail(_pipelineSnapTime)
  }

  /**
   * Defines the latest timestamp to be used for a give module as a TimeType
   * @param moduleID moduleID for which to get the until Time
   * @return
   */
  def untilTime(moduleID: Int): TimeTypes = {
    val startSecondPlusMaxDays = fromTime(moduleID).asLocalDateTime.plusDays(maxDays)
      .atZone(ZoneId.systemDefault()).toInstant.toEpochMilli
    val defaultUntilSecond = pipelineSnapTime.asUnixTimeMilli
    if (startSecondPlusMaxDays < defaultUntilSecond) {
      Config.createTimeDetail(startSecondPlusMaxDays)
    } else {
      Config.createTimeDetail(defaultUntilSecond)
    }
  }

  private[overwatch] def overwatchSchemaVersion: String = _overwatchSchemaVersion

  private[overwatch] def lastRunDetail: Array[ModuleStatusReport] = _lastRunDetail

  private[overwatch] def isLocalTesting: Boolean = _isLocalTesting

  private[overwatch] def isDBConnect: Boolean = _isDBConnect

  private[overwatch] def isFirstRun: Boolean = _isFirstRun

  private[overwatch] def debugFlag: Boolean = _debugFlag

  private[overwatch] def cloudProvider: String = _cloudProvider

  private[overwatch] def initialShuffleParts: Int = _intialShuffleParts

  private[overwatch] def maxDays: Int = _maxDays

  private[overwatch] def databaseName: String = _databaseName

  private[overwatch] def databaseLocation: String = _databaseLocation

  private[overwatch] def workspaceURL: String = _workspaceUrl

  private[overwatch] def token: String = cipher.decrypt(_token)

  private[overwatch] def apiEnv: ApiEnv = _apiEnv

  private[overwatch] def encryptedToken: Array[Byte] = _token

  private[overwatch] def auditLogConfig: AuditLogConfig = _auditLogConfig

  private[overwatch] def badRecordsPath: String = _badRecordsPath

  private[overwatch] def passthroughLogPath: Option[String] = _passthroughLogPath

  private[overwatch] def inputConfig: OverwatchParams = _inputConfig

  private[overwatch] def runID: String = _runID

  /**
   * OverwatchScope defines the modules active for the current run
   * Some have been disabled for the moment in a sprint to v1.0 release but this acts as the
   * cononical module inventory
   * @return
   */
  private[overwatch] def orderedOverwatchScope: Seq[OverwatchScope.Value] = {
    import OverwatchScope._
    //    jobs, clusters, clusterEvents, sparkEvents, pools, audit, passthrough, profiles
    Seq(audit, jobs, clusters, clusterEvents, sparkEvents, notebooks)
  }

  private[overwatch] def overwatchScope: Seq[OverwatchScope.Value] = _overwatchScope

  private[overwatch] def cal: Calendar = Calendar.getInstance(TimeZone.getTimeZone("UTC"))

  private[overwatch] def registerInitialSparkConf(value: Map[String, String]): this.type = {
    val manualOverrides = Map(
      "spark.databricks.delta.properties.defaults.autoOptimize.autoCompact" ->
        value.getOrElse("spark.databricks.delta.properties.defaults.autoOptimize.autoCompact", "false"),
      "spark.databricks.delta.properties.defaults.autoOptimize.optimizeWrite" ->
        value.getOrElse("spark.databricks.delta.properties.defaults.autoOptimize.optimizeWrite", "false"),
      "spark.databricks.delta.optimize.maxFileSize" ->
        value.getOrElse("spark.databricks.delta.optimize.maxFileSize", (1024 * 1024 * 128).toString),
      "spark.databricks.delta.retentionDurationCheck.enabled" ->
        value.getOrElse("spark.databricks.delta.retentionDurationCheck.enabled", "true"),
      "spark.sql.caseSensitive" -> "false"
    )
    _initialSparkConf = value ++ manualOverrides
    this
  }

  private[overwatch] def initialSparkConf: Map[String, String] = {
    _initialSparkConf
  }

  private[overwatch] def parsedConfig: ParsedConfig = {
    ParsedConfig(
      auditLogConfig = auditLogConfig,
      overwatchScope = overwatchScope.map(_.toString),
      tokenUsed = _tokenType,
      targetDatabase = databaseName,
      targetDatabaseLocation = databaseLocation,
      passthroughLogPath = passthroughLogPath
    )
  }

  /**
   * BEGIN SETTERS
   */

  /**
   * Snapshot time of the time the snapshot was started. This is used throughout the process as the until timestamp
   * such that every data point to be loaded during the current run must be < this pipeline SnapTime.
   * NOTE: PipelineSnapTime is EXCLUSIVE meaning < ONLY NOT <=
   * @return
   */
  private[overwatch] def setPipelineSnapTime(): this.type = {
    _pipelineSnapTime = LocalDateTime.now(ZoneId.of("Etc/UTC")).toInstant(ZoneOffset.UTC).toEpochMilli
    this
  }

  // TODO -- This is for local testing only
  /**
   * Test setter to simulate snapshot times. During a simulation where this is set, all data retrieved must be before
   * this time. This should not be a public-facing function in the final delivery
   * @param tsMilli long in milliseconds as per unix epoch
   * @return
   */
  def setPipelineSnapTime(tsMilli: Long): this.type = {
    //    _pipelineSnapTime = LocalDateTime.now(ZoneId.of("Etc/UTC")).toInstant(ZoneOffset.UTC).toEpochMilli
    _pipelineSnapTime = tsMilli
    this
  }

  /**
   * Identify the initial value before overwatch for shuffle partitions. This value gets modified a lot through
   * this process but should be set back to the same as the value before Overwatch process when Overwatch finishes
   * its work
   * @param value number of shuffle partitions to be set
   * @return
   */
  private[overwatch] def setInitialShuffleParts(value: Int): this.type = {
    _intialShuffleParts = value
    this
  }

  private[overwatch] def setMaxDays(value: Int): this.type = {
    _maxDays = value
    this
  }

  private[overwatch] def setOverwatchScope(value: Seq[OverwatchScope.Value]): this.type = {
    _overwatchScope = value
    this
  }

  def setDebugFlag(value: Boolean): this.type = {
    _debugFlag = value
    this
  }

  private[overwatch] def setIsFirstRun(value: Boolean): this.type = {
    _isFirstRun = value
    this
  }

  //TODO -- switch back to private -- public for testing only
  //private[overwatch]
  def setLastRunDetail(value: Array[ModuleStatusReport]): this.type = {
    // Todo -- Add assertion --> number of rows <= number of modules or something to that effect
    _lastRunDetail = value
    this
  }

  private def setApiEnv(value: ApiEnv): this.type = {
    _apiEnv = value
    this
  }

  /**
   * After the input parameters have been serialized into the OverwatchParams object, set it
   * @param value
   * @return
   */

  def setInputConfig(value: OverwatchParams): this.type = {
    _inputConfig = value
    this
  }

  // TODO - figure out why decryption doesn't consistently work and correc this function. The setters for
  //  rawToken should be removed throughout the codebase when this is resolved.
  /**
   * This function wraps all requisite parameters required for API calls and sets the complex type: APIEnv.
   * Also sets all required details to enable API calls. Furthermore, this function attempts
   * to encrypt the token and store as an attribute of the APIEnv.
   * An additional review should be completed to determine if all setters can be removed except the APIEnv registration.
   * Goal is to store a complex object, "APIEnv" with an encrypted token only.
   *
   * There have been challenges in getting this to work consistently and as of this
   * writing is not being used (the encrypted version). Notice the use of a "rawToken" which is fine but extra care
   * must be used to ensure the raw token doesn't get stored in clear text in any logs or passed in any
   * fashion that presents risk.
   * @param tokenSecret Optional input of the Token Secret. If left null, the token secret will be initialized
   *                    as the job owner or notebook user (if called from notebook)
   * @return
   */
  private[overwatch] def registeredEncryptedToken(tokenSecret: Option[TokenSecret]): this.type = {
    var rawToken = ""
    try {
      // Token secrets not supported in local testing
      if (tokenSecret.nonEmpty && !_isLocalTesting) { // not local testing and secret passed
        _workspaceUrl = dbutils.notebook.getContext().apiUrl.get
        _cloudProvider = if (_workspaceUrl.toLowerCase().contains("azure")) "azure" else "aws"
        val scope = tokenSecret.get.scope
        val key = tokenSecret.get.key
        rawToken = dbutils.secrets.get(scope, key)
        _token = cipher.encrypt(dbutils.secrets.get(scope, key))
        val authMessage = s"Valid Secret Identified: Executing with token located in secret, $scope : $key"
        logger.log(Level.INFO, authMessage)
        _tokenType = "Secret"
      } else {
        if (_isLocalTesting) { // Local testing env vars
          _workspaceUrl = System.getenv("OVERWATCH_ENV")
          //_cloudProvider setter not necessary -- done in local setup
          rawToken = System.getenv("OVERWATCH_TOKEN")
          _token = cipher.encrypt(System.getenv("OVERWATCH_TOKEN"))
          _tokenType = "Environment"
        } else { // Use default token for job owner
          _workspaceUrl = dbutils.notebook.getContext().apiUrl.get
          _cloudProvider = if (_workspaceUrl.toLowerCase().contains("azure")) "azure" else "aws"
          rawToken = dbutils.notebook.getContext().apiToken.get
          _token = cipher.encrypt(dbutils.notebook.getContext().apiToken.get)
          val authMessage = "No secret parameters provided: attempting to continue with job owner's token."
          logger.log(Level.WARN, authMessage)
          println(authMessage)
          _tokenType = "Owner"
        }
      }
      setApiEnv(ApiEnv(isLocalTesting, workspaceURL, rawToken, _token, cipher))
      this
    } catch {
      case e: Throwable => logger.log(Level.FATAL, "No valid credentials and/or Databricks URI", e); this
    }
  }

  /**
   * Set Overwatch DB and location
   * @param dbName
   * @param dbLocation
   * @return
   */
  private[overwatch] def setDatabaseNameandLoc(dbName: String, dbLocation: String): this.type = {
    _databaseLocation = dbLocation
    _databaseName = dbName
    println(s"DEBUG: Database Name and Locations set to ${_databaseName} and ${_databaseLocation}")
    this
  }

  /**
   * Sets audit log config
   * @param value
   * @return
   */
  private[overwatch] def setAuditLogConfig(value: AuditLogConfig): this.type = {
    _auditLogConfig = value
    this
  }

  /**
   * Sets Passthrough log path
   * Passthrough logs will be enabled post v1.0.
   * @param value
   * @return
   */
  private[overwatch] def setPassthroughLogPath(value: Option[String]): this.type = {
    _passthroughLogPath = value
    this
  }

  /**
   * Some input files have corrupted records, this sets the quarantine path for all such scenarios
   * @param value
   * @return
   */
  private[overwatch] def setBadRecordsPath(value: String): this.type = {
    _badRecordsPath = value
    this
  }

  /**
   * Manual setters for DB Remote and Local Testing. This is not used if "isLocalTesting" == false
   * This function allows for hard coded parameters for rapid integration testing and prototyping
   * @return
   */
  def buildLocalOverwatchParams(): String = {

    registeredEncryptedToken(None)
    _overwatchScope = Array(OverwatchScope.audit, OverwatchScope.clusters)
    val inputOverwatchScope = Array("audit", "clusters")
    _databaseName = "overwatch_local"
    _badRecordsPath = "/tmp/tomes/overwatch/sparkEventsBadrecords"
    //    _databaseLocation = "/Dev/git/Databricks--Overwatch/spark-warehouse/overwatch.db"

    // AWS TEST
    _cloudProvider = "azure"
    //    _auditLogConfig = AuditLogConfig(Some("/mnt/tomesdata/logs/field_training_audit/"), None)

    // AZURE TEST
    //    _cloudProvider = "azure"
    //    val EVENT_HUB_CONNECTION_STRING = "Endpoint=sb://testconsumerems.servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=AMtHCZQLecIxK51ZfZbrb/zhXdHZqrzrBbB+jDNgboQ="
    //    val EVENT_HUB_NAME = "overwatch"
    //    val MAX_EVENTS_PER_TRIGGER = 10000
    //    val STATE_PREFIX = "abfss://databricksoverwatch@hebdatalake05.dfs.core.windows.net/overwatch/auditlogs/state"
    //    val ehConfig = AzureAuditLogEventhubConfig(EVENT_HUB_CONNECTION_STRING, EVENT_HUB_NAME, STATE_PREFIX)
    //    _auditLogConfig = AuditLogConfig(None, Some(ehConfig))
    //
    //    OverwatchParams(
    //      _auditLogConfig, None, Some(DataTarget(Some(_databaseName), Some(_databaseLocation))),
    //      Some(_badRecordsPath), Some(inputOverwatchScope)
    //    )

    // FROM RAW PARAMS TEST
    """
      |{"auditLogConfig":{"azureAuditLogEventhubConfig":{"connectionString":"Endpoint=sb://testconsumerems.servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=AMtHCZQLecIxK51ZfZbrb/zhXdHZqrzrBbB+jDNgboQ=","eventHubName":"overwatch","auditRawEventsPrefix":"abfss://databricksoverwatch@hebdatalake05.dfs.core.windows.net/overwatch/auditlogs/state/overwatch","maxEventsPerTrigger":10000}},"tokenSecret":{"scope":"overwatch_test","key":"overwatch_updated"},"dataTarget":{"databaseName":"overwatch","databaseLocation":"dbfs:/user/hive/warehouse/overwatch.db"},"badRecordsPath":"/tmp/tomes/overwatch/sparkEventsBadrecords","overwatchScope":["audit","clusters","clusterEvents"],"maxDaysToLoad":60}
      |""".stripMargin

  }
}

object Config {
  val utcZone: ZoneId = ZoneId.of("Etc/UTC")

  /**
   * Most of Overwatch uses a custom time type, "TimeTypes" which simply pre-builds the most common forms / formats
   * of time. The sheer number of sources and heterogeneous time rules makes time management very challenging,
   * the idea here is to get it right once and just get the time type necessary.
   * @param tsMilli Unix epoch as a Long in milliseconds
   * @return
   */
  def createTimeDetail(tsMilli: Long): TimeTypes = {
    // TODO: systemDefault is the ticking mine
    val localDT = new Date(tsMilli).toInstant.atZone(ZoneId.systemDefault()).toLocalDateTime
    //    val localDT = LocalDateTime.ofInstant(dt.toInstant, ZoneId.of("Etc/UTC"))
    val instant = Instant.ofEpochMilli(tsMilli)
    TimeTypes(
      tsMilli, // asUnixTimeMilli
      lit(from_unixtime(lit(tsMilli).cast("double") / 1000).cast("timestamp")), // asColumnTS in local time,
      Date.from(instant), // asLocalDateTime at UTC
      instant.atZone(utcZone), // asUTCZonedDateTime
      localDT, // asLocalDateTime
      localDT.toLocalDate.atStartOfDay(ZoneId.systemDefault()).toInstant.toEpochMilli // asMidnightEpochMilli
    )
  }
}
